libsignal-service/src/cipher.rs

```diff
@@ -10,60 +22,58 @@ use crate::{
-    SessionCipher, StoreContext,
-};
-
-use block_modes::block_padding::{Iso7816, Padding};
-use prost::Message;
-
 /// Decrypts incoming messages and encrypts outgoing messages.
 ///
 /// Equivalent of SignalServiceCipher in Java.
-#[derive(Clone)]
-pub struct ServiceCipher {
-    pub(crate) context: Context,
-    pub(crate) store_context: StoreContext,
-    pub(crate) local_address: ServiceAddress,
-    sealed_session_cipher: SealedSessionCipher,
+pub struct ServiceCipher<R: Rng + CryptoRng + Clone> {
+    session_store: Box<dyn SessionStore>,
+    identity_key_store: Box<dyn IdentityKeyStore>,
+    signed_pre_key_store: Box<dyn SignedPreKeyStore>,
+    pre_key_store: Box<dyn PreKeyStore>,
+    csprng: R,
+    sealed_session_cipher: SealedSessionCipher<R>,
 }
 
-impl ServiceCipher {
-    pub fn from_context(
-        context: Context,
-        store_context: StoreContext,
-        local_address: ServiceAddress,
+impl<R: Rng + CryptoRng + Clone> ServiceCipher<R> {
+    pub fn new(
+        session_store: impl SessionStore + Clone + 'static,
+        identity_key_store: impl IdentityKeyStore + Clone + 'static,
+        signed_pre_key_store: impl SignedPreKeyStore + Clone + 'static,
+        pre_key_store: impl PreKeyStore + Clone + 'static,
         certificate_validator: CertificateValidator,
+        csprng: R,
     ) -> Self {
         Self {
-            context: context.clone(),
-            store_context: store_context.clone(),
-            local_address: local_address.clone(),
+            session_store: Box::new(session_store.clone()),
+            identity_key_store: Box::new(identity_key_store.clone()),
+            signed_pre_key_store: Box::new(signed_pre_key_store.clone()),
+            pre_key_store: Box::new(pre_key_store.clone()),
+            csprng: csprng.clone(),
             sealed_session_cipher: SealedSessionCipher::new(
-                context,
-                store_context,
-                local_address,
+                session_store,
+                identity_key_store,
+                signed_pre_key_store,
+                pre_key_store,
                 certificate_validator,
+                csprng,
             ),
         }
     }
 
     /// Opens ("decrypts") an envelope.
     ///
     /// Envelopes may be empty, in which case this method returns `Ok(None)`
-    pub fn open_envelope(
+    pub async fn open_envelope(
         &mut self,
         envelope: Envelope,
     ) -> Result<Option<Content>, ServiceError> {
         if envelope.legacy_message.is_some() {
-            let plaintext = self.decrypt(&envelope)?;
+            let plaintext = self.decrypt(&envelope).await?;
             let message =
                 crate::proto::DataMessage::decode(plaintext.data.as_slice())?;
             Ok(Some(Content::from_body(message, plaintext.metadata)))
         } else if envelope.content.is_some() {
-            let plaintext = self.decrypt(&envelope)?;
+            let plaintext = self.decrypt(&envelope).await?;
```

Comment:
FWIW, we need to discuss whether decrypting envelopes is still necessary. OWS stopped using the signaling key for encrypting envelopes, apparently.